File: packages/types/src/hermes-reasoning.ts
/**
 * Hermes TUI paints a `┌─ Reasoning ──┐` box (and a quiet-mode prelude
 * without the │ body). That decoration is presentation, not the reply —
 * Rivet Bot / RivetHub must never show it as assistant text.
 *
 * Keep the copy in integrations/hermes/rivet-den/hooks/hermes-den-hook.mjs
 * and apps/rivethub-android HermesReasoning.kt in sync with this.
 */

export interface HermesSplit {
  reasoning: string
  text: string
}

/** CSI + OSC so a colour-wrapped box still matches. */
export function stripAnsi(s: string): string {
  return (
    s
      // eslint-disable-next-line no-control-regex
      .replace(/\u001b\[[0-9;?]*[ -/]*[@-~]/g, '')
      // eslint-disable-next-line no-control-regex
      .replace(/\u001b\][^\u0007\u001b]*(?:\u0007|\u001b\\)/g, '')
  )
}

const HEADER = /^[\s]*[┌╭][─━\s]*\b(Reasoning|Thought|Thinking)\b/i
const FOOTER = /^[\s]*[└╰][─━\s]+[┘╯][\s]*$/
const BODY = /^[\s]*[│┃├┤]/

export function splitHermesReasoning(input: string): HermesSplit {
  if (!input) return { reasoning: '', text: '' }
  const lines = input.split(/\r?\n/)
  const reasoning: string[] = []
  const text: string[] = []
  let i = 0
  while (i < lines.length) {
    const vis = stripAnsi(lines[i])
    if (!HEADER.test(vis)) {
      text.push(lines[i])
      i += 1
      continue
    }
    const headerIdx = i
    i += 1
    let sawBox = false
    let hasTerminator = false
    while (i < lines.length) {
      const v = stripAnsi(lines[i])
      if (FOOTER.test(v)) {
        hasTerminator = true
        i += 1
        break
      }
      if (BODY.test(v)) {
        sawBox = true
        reasoning.push(v.replace(/^[\s]*[│┃├┤]\s?/, ''))
        i += 1
        continue
      }
      if (sawBox) {
        reasoning.push(v)
        i += 1
        continue
      }
      if (!v.trim()) {
        hasTerminator = true
        i += 1
        break
      }
      reasoning.push(v)
      i += 1
    }
    if (!sawBox && !hasTerminator) {
      text.push(lines[headerIdx])
      for (const line of reasoning) text.push(line)
      reasoning.length = 0
    }
  }
  return { reasoning: reasoning.join('\n').trim(), text: text.join('\n').trim() }
}

File: apps/rivet-bots-android/app/src/main/java/dev/rivetos/bots/data/HermesReasoning.kt
package dev.rivetos.bots.data

/**
 * Hermes TUI paints a `┌─ Reasoning ──┐` box into assistant text.
 * Keep in sync with packages/types/src/hermes-reasoning.ts.
 */
data class HermesSplit(val reasoning: String, val text: String)

private val ANSI = Regex("""\u001B\[[0-9;?]*[ -/]*[@-~]|\u001B\][^\u0007\u001B]*(?:\u0007|\u001B\\)""")
private val HEADER = Regex("""^[\s]*[┌╭][─━\s]*\b(Reasoning|Thought|Thinking)\b""", RegexOption.IGNORE_CASE)
private val FOOTER = Regex("""^[\s]*[└╰][─━\s]+[┘╯][\s]*$""")
private val BODY = Regex("""^[\s]*[│┃├┤]""")
private val BODY_PREFIX = Regex("""^[\s]*[│┃├┤]\s?""")

fun stripAnsi(s: String): String = ANSI.replace(s, "")

fun splitHermesReasoning(input: String): HermesSplit {
    if (input.isEmpty()) return HermesSplit("", "")
    val lines = input.split(Regex("""\r?\n"""))
    val reasoning = ArrayList<String>()
    val text = ArrayList<String>()
    var i = 0
    while (i < lines.size) {
        val vis = stripAnsi(lines[i])
        if (!HEADER.containsMatchIn(vis)) {
            text.add(lines[i])
            i++
            continue
        }
        val headerIdx = i
        i++
        var sawBox = false
        var hasTerminator = false
        while (i < lines.size) {
            val v = stripAnsi(lines[i])
            if (FOOTER.matches(v)) {
                hasTerminator = true
                i++
                break
            }
            if (BODY.containsMatchIn(v)) {
                sawBox = true
                reasoning.add(BODY_PREFIX.replaceFirst(v, ""))
                i++
                continue
            }
            if (sawBox) {
                reasoning.add(v)
                i++
                continue
            }
            if (v.isBlank()) {
                hasTerminator = true
                i++
                break
            }
            reasoning.add(v)
            i++
        }
        if (!sawBox && !hasTerminator) {
            text.add(lines[headerIdx])
            text.addAll(reasoning)
            reasoning.clear()
        }
    }
    return HermesSplit(reasoning.joinToString("\n").trim(), text.joinToString("\n").trim())
}

fun visibleAssistantText(text: String): String = splitHermesReasoning(text).text
